### DISCUSSION TITLE
What does bundle provides?

### DISCUSSION THREAD

**Discussion by liverpool1026** [sqlalchemy/sqlalchemy#8484] [Usage Questions] [answered]
Hi there, I was reading the docs and came across Bundle object, I don't think I fully understand what it does, if I can get some clarification that would be appreciated.

My understanding is the bundle can provide a different view.
```
from sqlalchemy.orm import Bundle

bn = Bundle('mybundle', MyClass.data1, MyClass.data2)
for row in session.query(bn).filter(bn.c.data1 == 'd1'):
    print(row.mybundle.data1, row.mybundle.data2)
```

So now both data1 and data2 are under the mybundle namespace

It can also be subclassed to provide custom behaviours when result are fetched.

It was also mentioned in the section regarding performance.
https://docs.sqlalchemy.org/en/14/faq/performance.html#result-fetching-slowness-orm

How does using bundle affect performance?

**Comment by zzzeek** [comment, marked as answer]
> Hi there, I was reading the docs and came across Bundle object, I don't think I fully understand what it does, if I can get some clarification that would be appreciated.
> 
> My understanding is the bundle can provide a different view.

that's correct

> 
> ```
> from sqlalchemy.orm import Bundle
> 
> bn = Bundle('mybundle', MyClass.data1, MyClass.data2)
> for row in session.query(bn).filter(bn.c.data1 == 'd1'):
>     print(row.mybundle.data1, row.mybundle.data2)
> ```
> 
> So now both data1 and data2 are under the mybundle namespace

correct!

> 
> It can also be subclassed to provide custom behaviours when result are fetched.

yup, simple things like a dictionary

> 
> It was also mentioned in the section regarding performance. https://docs.sqlalchemy.org/en/14/faq/performance.html#result-fetching-slowness-orm
> 
> How does using bundle affect performance?

fetching Bundle is faster than fetching a full ORM object.  Like if you were loading MyModel objects, and you just wanted a small set of plain columns from each MyModel for read only, you could instead put those columns in a Bundle and the ORM will have much less work to do in order to assemble Bundle objects vs. full MyModel objects, as there is no identity map lookup, the Bundle returns lightweight named tuples without a `__dict__`, etc.

This is demonstrated at https://docs.sqlalchemy.org/en/14/_modules/examples/performance/large_resultsets.html 

demo of two comparable tests, which you can run from the SQLAlchemy source checkout:

```py
[classic@photon3 sqlalchemy:rel_1_4]$ python -m examples.performance.large_resultsets --test  test_orm_full_objects_list
Running setup once...
Tests to run: test_orm_full_objects_list
test_orm_full_objects_list : Load fully tracked ORM objects into one big list(). (500000 iterations); total time 5.635296 sec
[classic@photon3 sqlalchemy:rel_1_4]$ python -m examples.performance.large_resultsets --test  test_orm_bundles
Running setup once...
Tests to run: test_orm_bundles
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (500000 iterations); total time 2.617927 sec
```

**Reply by liverpool1026** [reply]
i do see the benefit of using bundle compared to querying the entire orm obj. But if I just need a subset of columns, doesn't query the columns individually performs better? (According to the examples)
```
(sqlalchemy) machine:~/Git/sqlalchemy$ python3 -m examples.performance.large_resultsets --test test_orm_bundles
Running setup once...
Tests to run: test_orm_bundles
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (500000 iterations); total time 1.010855 sec
(sqlalchemy) machine:~/Git/sqlalchemy$ python3 -m examples.performance.large_resultsets --test test_orm_columns
Running setup once...
Tests to run: test_orm_columns
test_orm_columns : Load individual columns into named tuples using the ORM. (500000 iterations); total time 0.750711 sec
```

**Reply by zzzeek** [reply]
yes a subset of columns performs better.   Bundle applies that additional grouping that may be desirable in some applications.

**Reply by liverpool1026** [reply]
Thanks
